### PR TITLE
GH#1716: optimize front-end My Sites toolbar

### LINKED ISSUE #1716: Add an opt-in optimized front-end My Sites toolbar
## Objective

Add an opt-in Ultimate Multisite setting that reduces authenticated front-end toolbar overhead on large networks by limiting the WordPress **My Sites** dropdown to the current site.

The complete site list must remain available from WordPress's **My Sites** administration page and all wp-admin/network-admin screens.

## Evidence

Production-scale local profiling used 114 network sites and a super administrator whose toolbar exposed 26 sites.

- Authenticated front-end baseline: **31,182 object-cache lookups**.
- Restricting the toolbar fixture to one visible site: **25,756 lookups**.
- Reduction: **5,426 lookups (17.4%)** for one request.
- `switch_blog` tracing recorded **160 events**. Two toolbar-related paths each produced 26 switch/restore pairs:
  - `WP_Ultimo\Managers\Site_Manager::hide_customer_sites_from_super_admin_list()`
  - WordPress core `wp_admin_bar_my_sites_menu()`
- Every switch initializes active Polylang and WooCommerce integrations for that blog, amplifying options, network-options, users, terms, and translation cache traffic.

The issue is not the database query that selects sites. It is eagerly preparing and rendering every site in the authenticated front-end toolbar.

## Reproducer

1. Configure a multisite network with at least 25 sites visible to a super administrator.
2. Enable Ultimate Multisite and log in as that super administrator.
3. Request a front-end page with the WordPress toolbar enabled.
4. Trace `switch_blog`/`restore_current_blog` and object-cache lookups for the request.
5. Observe one switch/restore pair per visible site from `hide_customer_sites_from_super_admin_list()` and another pair per site from `wp_admin_bar_my_sites_menu()`.
6. Repeat with only the current site exposed to the toolbar; compare switch and cache-read counts.

**Symptom command** (run with an authenticated cookie against the local multisite fixture while request-scoped cache and `switch_blog` tracers are active):

```bash
curl -sS -H 'Cookie: <authenticated-super-admin-cookie>' https://<local-multisite-host>/ > /dev/null
```

**Actual output**:

```text
object-cache lookups: 31,182
switch_blog events: 160
Site_Manager toolbar-related pairs: 26
WordPress My Sites menu pairs: 26
```

**Expected output**:

```text
When the opt-in setting is enabled, front-end toolbar preparation is bounded to the current site and does not add one switch/restore pair per visible network site. Administrative My Sites listings remain complete.
```

**Causal status**: confirmed

**Owning-path proof**:

- **Production entry point**: `WP_Admin_Bar::initialize()` calls `get_blogs_of_user()` while preparing an authenticated toolbar.
- **Call chain**: `get_blogs_of_user()` → `pre_get_blogs_of_user` → `Site_Manager::hide_customer_sites_from_super_admin_list()` → per-site details; core later runs `wp_admin_bar_my_sites_menu()` → `switch_to_blog()` for every returned site.
- **Integrated verification**: the same authenticated front-end request fell from 31,182 to 25,756 object-cache lookups when the toolbar fixture exposed only the current site.

## Proposed setting

Add an opt-in toggle under **Ultimate Multisite → Settings → Sites → Site Options**:

> **Optimize the front-end My Sites toolbar**
>
> Show only the current site in the front-end toolbar to reduce processing on large networks. The complete site list remains available from the My Sites administration page.

Suggested setting key: `optimize_frontend_my_sites_toolbar`.

- Default to `false` for backward compatibility.
- Apply only to authenticated front-end requests where the toolbar is displayed.
- Keep existing behavior unchanged in site admin, network admin, user admin, REST, AJAX, cron, and WP-CLI requests.
- Preserve the current-site Dashboard/Visit Site links, the top-level My Sites link, and Network Admin links.
- Do not alter membership, site ownership, or capability results outside the toolbar optimization path.

## Implementation context

Expected files:

- `inc/class-settings.php`
  - Register the toggle in the existing `sites` section near the Site Options fields.
  - Follow existing `add_field(..., ['type' => 'toggle', 'default' => 0])` patterns.
- `inc/managers/class-site-manager.php`
  - `init()` registers `hide_customer_sites_from_super_admin_list()` on `pre_get_blogs_of_user` at priority 999.
  - Add a narrowly scoped fast path when the setting is enabled and the request is the authenticated front-end toolbar path.
  - Return only the current/active site needed by `WP_Admin_Bar`; avoid reading `blogname`/`siteurl` details for every eligible site.
  - Leave the existing full-list path untouched in administrative and non-toolbar contexts.
- `tests/WP_Ultimo/Managers/Site_Manager_Test.php`
  - Extend the existing tests around `hide_customer_sites_from_super_admin_list()`.
- `tests/WP_Ultimo/Settings_Test.php`
  - Verify registration and the backward-compatible default.

If safely distinguishing the toolbar's nested `get_blogs_of_user()` call requires a small dedicated helper or lifecycle flag, keep that implementation internal to Ultimate Multisite and document why the scope cannot rely on `! is_admin()` alone.

## Acceptance criteria

- [ ] The new setting exists, is documented in its field description, and defaults off.
- [ ] With the setting off, the existing site list and toolbar behavior are unchanged.
- [ ] With the setting on, an authenticated front-end toolbar request prepares only the current site.
- [ ] The full site list remains available in `wp-admin/my-sites.php`, site admin, and network admin.
- [ ] Non-toolbar calls to `get_blogs_of_user()` retain complete, correctly filtered results.
- [ ] Super-admin filtering and customer-owned-site hiding remain correct.
- [ ] The toolbar retains current-site, My Sites, and Network Admin navigation.
- [ ] Tests cover enabled/disabled, front-end/admin, toolbar/non-toolbar, super-admin/non-super-admin, and multisite edge cases.
- [ ] A performance regression test or bounded instrumentation demonstrates that toolbar work no longer scales linearly with the number of visible sites when enabled.

## Verification

Run:

```bash
vendor/bin/phpcs inc/class-settings.php inc/managers/class-site-manager.php tests/WP_Ultimo/Managers/Site_Manager_Test.php tests/WP_Ultimo/Settings_Test.php
vendor/bin/phpunit --filter Site_Manager_Test
vendor/bin/phpunit --filter Settings_Test
```

Then verify in a multisite fixture with at least 25 visible sites:

1. Setting off: full front-end toolbar site list remains present.
2. Setting on: front-end toolbar shows only the current site plus My Sites/Network Admin navigation.
3. Setting on: `wp-admin/my-sites.php` and network-admin site listings remain complete.
4. Compare `switch_blog` counts before and after; the optimized front-end request should not perform one switch/restore pair per visible toolbar site.

## Out of scope

- Changing WordPress core toolbar behavior globally.
- Removing the toolbar entirely.
- Altering site ownership or membership visibility.
- Optimizing unrelated Polylang or WooCommerce callbacks.

### PR DESCRIPTION
**Pull request by superdav42**
## Summary

- Add an opt-in setting to limit front-end My Sites toolbar preparation to the active site.
- Preserve complete My Sites and administration listings by requiring the WordPress admin-bar initialization call path.
- Cover the setting default, toolbar scope, and one-site result.

## Verification

- `vendor/bin/phpcs inc/class-settings.php inc/managers/class-site-manager.php tests/WP_Ultimo/Managers/Site_Manager_Test.php tests/WP_Ultimo/Settings_Test.php`
- `vendor/bin/phpunit --filter test_frontend_my_sites_toolbar_optimization --debug`
- `vendor/bin/phpunit --filter Settings_Test`

## Known verification limitation

`vendor/bin/phpunit --filter Site_Manager_Test` stops when an existing test accesses unavailable database tables; the three new focused Site_Manager tests pass.

Resolves #1716

## Summary by CodeRabbit

* **New Features**
  * Added an optional optimization for the front-end “My Sites” toolbar.
  * When enabled, eligible front-end requests display only the current site, including for non-super-admin users.
  * Archived, spam, or deleted sites are excluded when the full site list is not requested.
  * Other contexts continue using the existing site list behavior.

* **Configuration**
  * Added a toggle for the optimization.
  * The setting is disabled by default.

### CODE REVIEW

**Comment by superdav42** [comment]
## Merge summary

- Adds an opt-in setting that limits front-end My Sites toolbar preparation to the current site.
- Scopes the fast path to `WP_Admin_Bar::initialize()` so wp-admin, REST, CLI, AJAX, cron, and ordinary `get_blogs_of_user()` calls keep their existing behavior.
- Adds setting and focused manager coverage.

## Verification

- Passed: `vendor/bin/phpcs inc/class-settings.php inc/managers/class-site-manager.php tests/WP_Ultimo/Managers/Site_Manager_Test.php tests/WP_Ultimo/Settings_Test.php` (existing warnings only)
- Passed: `vendor/bin/phpunit --filter test_frontend_my_sites_toolbar_optimization --debug`
- Passed: `vendor/bin/phpunit --filter Settings_Test`
- Environment limitation: the full `Site_Manager_Test` filter stops on unavailable database tables outside this change; focused new tests pass.

---
Merged via PR #1719 to main.
_Merged by deterministic merge pass (pulse-wrapper.sh)._

---
[aidevops.sh](https://aidevops.sh) v3.32.317 plugin for [OpenCode](https://opencode.ai) v1.18.29 spent 1m on this as a headless bash routine.

File: inc/managers/class-site-manager.php [hunk 1]
```diff
@@ -1163,6 +1163,10 @@ public function hide_customer_sites_from_super_admin_list($sites, $user_id, $all
 			return $sites;
 		}
 
+		if ($this->is_frontend_my_sites_toolbar_request()) {
+			return $this->get_current_site_for_frontend_my_sites_toolbar();
+		}
```
[the remark this answers is not in the document]
Comment (superdav42) [reply inc/managers/class-site-manager.php:1170]: @coderabbitai Addressed in 7f697cbb. The optimized branch now applies `get_blogs_of_user` before returning, matching the standard path at `inc/managers/class-site-manager.php:1166`. Added a regression test in `tests/WP_Ultimo/Managers/Site_Manager_Test.php:2974`.

Verified with `vendor/bin/phpunit --filter test_frontend_my_sites_toolbar_optimization_applies_get_blogs_of_user_filter` and PHPCS; the commit hook also passed PHPStan.

**Comment by superdav42** [comment]
DISPATCH_CLAIM nonce=e5945845d88553e4811ed07b5b77ff07 runner=superdav42 ts=2026-09-09T13:10:43Z max_age_s=120 version=3.32.317 opencode_version=1.18.30 lease_token=e5945845d88553e4811ed07b5b77ff07 device=device-1783824528-2609248-26808 session=issue-1719 phase=prelaunch expires_at=1788959565

**Comment by superdav42** [comment]
REVIEW_FOLLOWUP_CREATED source_pr=1719 issue=1816 fingerprint=source-pr-1719 runner=superdav42 ts=2026-09-09T13:11:07Z
